No chords in pipeline

bartdegoede opened the pull request:
With the introduction of the collection refresh mechanism (see #77) a Celery chord construct was created in the pipeline, in order to determine when a pipeline has finished so the Elasticsearch aliases could be transferred to the newly created index.

However, this requires Redis to first create all parallel tasks and keep them in memory before the chord callback method updating the index alias could be called. That means that an extractor has to finish its run before any tasks could start being executed. This is undesirable, since for many collections (such as Rijksmuseum) it can take a long time to generate all tasks, and even worse, do not fit into memory.

This pull request provides a solution that generates a unique ID for each pipeline that is created, and stores that in the Celery result backend (currently we use Redis, but the implementation in `ocd_backend/result_backends.py` provides a simple interface to extend to other backends as well). Then, it creates an ID for each chain/item that is to be processed, and stores those in a set in the backend as well.

`ocd_backend/result_backends.py` provides a mixin for tasks that can remove the identifier of that chain from that set when an error occurs. The loaders in `ocd_backend/loaders/__init__.py` provide an additional hook that checks if there are any tasks currently left running (i.e. there are still values in the set), or if the extractor is still generating tasks.

**N.B.:** some additional unit tests are in the works, but this issue is blocking work on other components. @justinvw please review the changes carefully :-)

justinvw commented:
I did a first round of testing, some issues I discovered:
- In line 17 of ``pipeline.py`` you try the get the value of ``index_name``. However, the ``sources.json`` doesn't contain this value. I assumed you meant to use the ``id`` here, which is what I changed it to in order to continue with testing the PR.
- If an extractor throws an exception that we don't catch - which happens quite often with our 'stupid' extractors and the wobbly external services they have to hit - the extraction process stops but the ``run_identifier`` stored in Redis still remains. This means that the pipeline never 'finishes' and thus the cleanup in BaseLoader.after_return is never executed.

  Maybe wrap the ``extractor.run()`` loop in a ``try/except`` to cleanup the ``run_identifier`` if an exception occurs? If we implement it like this, we should make sure to re-raise the exception to indicate to the user that something went wrong with the extraction process.
- What happens when an exception occurs in the transform task in the last chain of a pipeline run? Because the cardinality of ``run_identifier_chains`` is then zero, and the ``run_identifier`` will no longer be present, to me it looks like the pipeline finishes without updating the alias (the code in ``ElasticsearchLoader.run_finished``).

  Maybe this is also the reason why you needed the cardinality check at the end of ``pipeline.py``. I think we should look into those last calls, by the way. It shouldn't be necessary.
- When the ES index doesn't already exist and I start the extractor, I see it appearing with the correct alias, documents get also added without any problems. However, when the tasks finish the index disappears...
- Some PEP8, mainly in ``pipeline.py`` :wink: .

bartdegoede commented:
I've put my answers in the same sequence of your issues, so it's a bit easier to relate them :-):
- `index_name` was introduced with the introduction of versioning indices in #77, so that we could have a collection name separate from the internal id. It now falls back to the collection id if no `index_name` is provided, and I've added names to `sources.json`
- Good point, I missed that (rather obvious) case. Implemented that as well. I've introduced another state to the identifier (which was "exist" or "does not exist"), as cleaning up the `run_identifier` could cleanup the key, while there were still chains running; this caused the recreation of the index that was just removed, and the task thinking it had completed successfully, causing it to update the alias, leaving me with indices with just a few documents. The `run_identifier` now either contains a value of `running`, `error` or `done`.
- I'm giving this one some more thought.
- Good catch, fixed.
- I bow my head in shame.

justinvw commented:
It looks like we've covered all the edge cases by now, so let's merge!
